File: android/app/src/main/kotlin/com/westinx/yurunavi/NativeLogBridge.kt
package com.westinx.yurunavi

import android.os.Handler
import android.os.Looper
import android.os.Process
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import java.io.BufferedReader
import java.io.InputStreamReader

/**
 * 네이티브(Kotlin/Flutter 엔진) 프로세스 로그를 EventChannel로 Dart에 흘려보낸다.
 *
 * 배경: 실주행 중 화면 서브트리가 백지로 렌더되는 버그의 유력 원인은 Flutter 엔진
 * (Impeller/Vulkan)의 GPU 자원 할당 실패인데, 기존 FileLogger는 Dart debugPrint만
 * 가로채 네이티브 엔진 로그(logcat)는 전혀 남기지 못했다. 이 브리지는 자기 프로세스
 * (자기 pid)의 logcat만 백그라운드 데몬 스레드로 읽어 allowlist 키워드에 매치되는
 * 줄만 Dart로 전달한다.
 *
 * 자기 pid 로그는 READ_LOGS 권한 없이도 읽을 수 있다(추가 권한/매니페스트 변경 없음).
 * 이 계측은 debug 전용이 아니다 — 마스터의 release 실주행에서 캡처돼야 의미가
 * 있으므로 kDebugMode 게이팅을 하지 않는다(Dart 쪽도 마찬가지).
 */
class NativeLogBridge(messenger: BinaryMessenger) {

    companion object {
        private const val CHANNEL = "com.westinx.yurunavi/native_log"

        /** 이 중 하나라도 포함된 로그 줄만 통과시킨다. `flutter` 태그 전체 허용은 절대 금지
         *  (debugPrint → logcat → 이 펌프 → EventChannel → debugPrint 무한 증폭 루프가 된다). */
        private val KEYWORDS = listOf(
            "Impeller", "impeller", "Vulkan", "vulkan", "glyph", "atlas",
            "GL_OUT_OF_MEMORY", "OutOfMemory", "onTrimMemory", "lowmemorykiller",
            "Surface", "EGL"
        )
    }

    private val mainHandler = Handler(Looper.getMainLooper())
    private var eventSink: EventChannel.EventSink? = null

    init {
        EventChannel(messenger, CHANNEL).setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
            }

            override fun onCancel(arguments: Any?) {
                eventSink = null
            }
        })
        startLogPump()
    }

    /** 앱 프로세스 종료 시 자연 종료되는 데몬 스레드 하나로 충분 — 정지/재개 관리는 하지 않는다. */
    private fun startLogPump() {
        val thread = Thread({
            try {
                val pid = Process.myPid()
                val process = ProcessBuilder("logcat", "-v", "threadtime", "--pid=$pid")
                    .redirectErrorStream(true)
                    .start()
                val reader = BufferedReader(InputStreamReader(process.inputStream))
                while (true) {
                    val line = reader.readLine() ?: break
                    if (KEYWORDS.any { line.contains(it) }) {
                        val captured = line
                        mainHandler.post { eventSink?.success(captured) }
                    }
                }
            } catch (e: Exception) {
                // 계측 실패가 앱을 죽이면 안 된다 — 조용히 종료
            }
        }, "ynav-native-log-pump")
        thread.isDaemon = true
        thread.start()
    }
}

File: lib/core/logging/native_log_bridge.dart
import 'package:flutter/services.dart';

import 'file_logger.dart';

/// `com.westinx.yurunavi/native_log` EventChannel(네이티브 `NativeLogBridge.kt`)을
/// 구독해 받은 줄마다 [FileLogger.writeRaw]로 흘려보낸다.
///
/// 배경: 실주행 중 화면 서브트리가 백지로 렌더되는 버그의 유력 원인은 Flutter
/// 엔진(Impeller/Vulkan) GPU 자원 할당 실패인데, 기존 [FileLogger]는 Dart
/// `debugPrint`만 가로채 네이티브 엔진 로그는 전혀 남기지 못했다. 이 구멍을
/// 메우는 계측 브리지 — debug 전용이 아니다.
///
/// [start]는 반드시 `FileLogger.init()` 이후에 호출해야 한다(그래야 `_sink`가
/// 준비돼 있다). 채널이 없거나 에러가 나도 앱이 죽으면 안 되므로 전부
/// try/catch로 감싼다.
class NativeLogBridge {
  static const EventChannel _channel =
      EventChannel('com.westinx.yurunavi/native_log');

  static void start() {
    try {
      // 앱 생애주기 전체에 걸쳐 계속 수신 — 취소 진입점이 없어도 구독 자체는
      // 채널 내부에서 유지된다. 반환값을 들고 있을 필요가 없다.
      _channel.receiveBroadcastStream().listen(
        (event) {
          try {
            if (event is String) {
              FileLogger.writeRaw(event);
            }
          } catch (_) {
            // 계측 실패가 앱을 죽이면 안 된다
          }
        },
        onError: (_) {},
        cancelOnError: false,
      );
    } catch (_) {
      // 채널이 없거나(구버전 네이티브) 구독 자체가 실패해도 무시하고 계속 진행
    }
  }
}
